Mathlib/Algebra/Lie/Weights/RootSystem.lean

```diff
@@ -0,0 +1,248 @@
+  rw [Weight.coe_neg, ← smul_neg, neg_neg, ← Weight.coe_neg, H₂] at this
+  exact this (weightSpace_chainTopCoeff_add_one_nsmul_add α β hα)
+
+lemma chainTopCoeff_le_chainLength : chainTopCoeff α β ≤ chainLength α β := by
+  by_cases hα : α.IsZero
+  · simp only [hα.eq, chainTopCoeff_zero, zero_le]
+  rw [← not_lt, ← Nat.succ_le]
+  intro e
+  apply weightSpace_nsmul_add_ne_bot_of_le α β
+    (Nat.sub_le (chainTopCoeff α β) (chainLength α β).succ)
+  rw [nsmul_eq_smul_cast ℤ, Nat.cast_sub e, sub_smul, sub_eq_neg_add,
+    add_assoc, ← coe_chainTop, ← nsmul_eq_smul_cast]
+  exact rootSpace_neg_nsmul_add_chainTop_of_lt α β hα (Nat.lt_succ_self _)
+
+lemma chainBotCoeff_add_chainTopCoeff :
+    chainBotCoeff α β + chainTopCoeff α β = chainLength α β := by
+  by_cases hα : α.IsZero
+  · rw [hα.eq, chainTopCoeff_zero, chainBotCoeff_zero, zero_add, chainLength_of_isZero α β hα]
+  apply le_antisymm
+  · rw [← Nat.le_sub_iff_add_le (chainTopCoeff_le_chainLength α β),
+      ← not_lt, ← Nat.succ_le, chainBotCoeff, ← Weight.coe_neg]
+    intro e
+    apply weightSpace_nsmul_add_ne_bot_of_le _ _ e
+    rw [nsmul_eq_smul_cast ℤ, Nat.cast_succ, Nat.cast_sub (chainTopCoeff_le_chainLength α β),
+      LieModule.Weight.coe_neg, smul_neg, ← neg_smul, neg_add_rev, neg_sub, sub_eq_neg_add,
+      ← add_assoc, ← neg_add_rev, add_smul, add_assoc, ← coe_chainTop, neg_smul,
+      ← @Nat.cast_one ℤ, ← Nat.cast_add, ← nsmul_eq_smul_cast]
+    exact rootSpace_neg_nsmul_add_chainTop_of_lt α β hα (Nat.lt_succ_self _)
+  · rw [← not_lt]
+    intro e
+    apply rootSpace_neg_nsmul_add_chainTop_of_le α β e
+    rw [← Nat.succ_add, nsmul_eq_smul_cast ℤ, ← neg_smul, coe_chainTop, ← add_assoc, ← add_smul,
+      Nat.cast_add, neg_add, add_assoc, neg_add_self, add_zero, neg_smul, ← smul_neg,
+      ← nsmul_eq_smul_cast]
+    exact weightSpace_chainTopCoeff_add_one_nsmul_add (-α) β (Weight.IsNonZero.neg hα)
+
+@[simp]
+lemma chainLength_neg :
+    chainLength (-α) β = chainLength α β := by
+  rw [← chainBotCoeff_add_chainTopCoeff, ← chainBotCoeff_add_chainTopCoeff, add_comm]
+  simp only [Weight.coe_neg, chainTopCoeff_neg, chainBotCoeff_neg]
+
+lemma chainTopCoeff_add_chainBotCoeff :
+    chainTopCoeff α β + chainBotCoeff α β = chainLength α β := by
+  rw [add_comm, chainBotCoeff_add_chainTopCoeff]
+
+lemma chainLength_sub_chainTopCoeff :
+    chainLength α β - chainTopCoeff α β = chainBotCoeff α β :=
+  Nat.sub_eq_of_eq_add (chainBotCoeff_add_chainTopCoeff α β).symm
+
+lemma chainLength_sub_chainBotCoeff :
+    chainLength α β - chainBotCoeff α β = chainTopCoeff α β :=
+  Nat.sub_eq_of_eq_add (chainTopCoeff_add_chainBotCoeff α β).symm
+
+lemma chainBotCoeff_le_chainLength : chainBotCoeff α β ≤ chainLength α β :=
+  (Nat.le_add_left _ _).trans_eq (chainTopCoeff_add_chainBotCoeff α β)
+
+/-- If `β - qα ... β ... β + rα` is the `α`-chain through `β`, then
+  `β (coroot α) = q - r`. In particular, it is an integer. -/
+lemma apply_coroot :
```

Comment:
The two names here should be very similar to aid discoverability.
```suggestion
lemma apply_coroot_eq_cast' :
```
Possibly the `'` should be on the other one but I don't mind which way it is.